packages/dds/tree/src/test/simple-tree/treeApi.spec.ts

```diff
@@ -108,56 +108,270 @@ describe("treeApi", () => {
 	});
 
 	describe("on", () => {
-		const sb = new SchemaFactory("object");
-		const object = sb.object("child", {
-			content: sb.number,
-		});
-		const list = sb.array(object);
-		const treeSchema = sb.object("parent", { object, list });
 
-		describe("events", () => {
-			function check(mutate: (root: NodeFromSchema<typeof treeSchema>) => void) {
-				it(".on(..) must subscribe to change event", () => {
+		describe("object node", () => {
+			const sb = new SchemaFactory("object-node-in-root");
+			class myObject extends sb.object("object", {
+				myNumber: sb.number,
+			}) {}
+			const treeSchema = sb.object("root", {
+				rootObject: myObject,
+			});
+
+			function check(
+				eventName: keyof TreeChangeEvents,
+				mutate: (root: NodeFromSchema<typeof treeSchema>) => void,
+			) {
+				it(`.on('${eventName}') subscribes and unsubscribes correctly`, () => {
 					const root = hydrate(treeSchema, {
-						object: { content: 1 },
-						list: [{ content: 2 }, { content: 3 }],
+						rootObject: {
+							myNumber: 1,
+						},
 					});
 					const log: any[][] = [];
 
-					Tree.on(root as TreeNode, "afterChange", (...args: any[]) => {
+					const unsubscribe = Tree.on(root, eventName, (...args: any[]) => {
 						log.push(args);
 					});
 
 					mutate(root);
 
-					const numChanges = log.length;
-					assert(
-						numChanges > 0,
-						"Must receive change notifications after subscribing to event.",
+					assert.equal(
+						log.length,
+						1,
+						`Must receive change notifications after subscribing to event '${eventName}'.`,
+					);
+
+					unsubscribe();
+
+					mutate(root);
+
+					assert.equal(
+						log.length,
+						1,
+						`Mutation after unsubscribe must not emit event '${eventName}'.`,
 					);
 				});
+			}
 
-				it(".on(..) must return unsubscribe function", () => {
-					const root = hydrate(treeSchema, {
-						object: { content: 1 },
-						list: [{ content: 2 }, { content: 3 }],
-					});
-					const log: any[][] = [];
+			check(
+				"afterShallowChange",
+				(root) =>
+					root.rootObject = new myObject({
+						myNumber: 2,
+					}),
+			);
+			check("afterDeepChange", (root) => root.rootObject.myNumber++);
+
+			it(`change to direct fields triggers both 'afterShallowChange' and 'afterDeepChange'`, () => {
```

Comment:
Every node kind has a pair of tests like this one and the one right after. I tried DRY-ing the code but I couldn't figure out the correct typing needed to pass around schemas and functions that depend on them.